routing/missioncontrol_store.go

```diff
@@ -268,8 +273,17 @@ func deserializeResult(k, v []byte) (*paymentResult, error) {
 // AddResult adds a new result to the db.
 func (b *missionControlStore) AddResult(rp *paymentResult) {
 	b.queueMx.Lock()
-	defer b.queueMx.Unlock()
 	b.queue.PushBack(rp)
+	signalRun := b.queue.Len() == 1
+	b.queueMx.Unlock()
+
+	// Signal run() that the queue is non-empty.
+	if signalRun {
+		select {
+		case <-b.done:
+		case b.queueChan <- struct{}{}:
+		}
+	}
```

Comment:
`signalRun` is determined under the `queueMx` lock, which also gates the emptying of `b.queue`, therefore it will only ever be true if the `b.queue` was previously empty.

If this method (`AddResult`) is called in quick succession, one of two things can happen:

- Either `storeResults()` already started and emptied `b.queue`, in which case this will start a new batch and a new timer
- Or it has not started yet, and the new result will be put and processed in the outstanding batch

If this is called in quick succession with an empty `b.queue`, only the first one (under the `m.queueMx` lock) will send the signal.